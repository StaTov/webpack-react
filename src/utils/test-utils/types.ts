import { RenderOptions } from "@testing-library/react"
import { RootState, AppStore } from "src/store/types"

export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>
    store?: AppStore
  }