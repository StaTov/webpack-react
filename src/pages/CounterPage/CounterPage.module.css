.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 10px;
  margin: 40px;
}
.value, .rangeValue {
  font-size: 70px;
  font-weight: 700;
  color: rgb(7, 108, 114);
}

button {
  padding: 12px;
  background-color: black;
  color: white;
  font-size: 20px;
  border: none;
  border-radius: 30px;
  cursor: pointer;
  transition: color 0.2s;
}

button:hover {
    opacity: 0.9;
}

button:active {
  transform: translateY(1px);
}

