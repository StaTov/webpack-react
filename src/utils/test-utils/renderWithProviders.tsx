import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { setupStore } from 'src/store'
import { ExtendedRenderOptions } from './types'
import { MemoryRouter } from 'react-router-dom'

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}