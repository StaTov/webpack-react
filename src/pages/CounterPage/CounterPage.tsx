import { useState } from "react";
import { useAppDispatch, useAppSelector } from "src/store/hooks";
import { getCounterValue } from "src/store/Selectors/counterSelector";
import { decrement, increment, incrementByAmount } from "src/store/Slices/counterSlice/counterSlice";
import { DECREMENT, INCREMENT, INCREMENT_BY_AMOUNT } from "src/utils/constants/CounterPage/buttonNames";
import style from "./CounterPage.module.css";

const CounterPage = () => {

    const dispatch = useAppDispatch()
    const counterValue = useAppSelector(getCounterValue)
    const [value, setValue] = useState(0)

    const handleIncrement = () => dispatch(increment());
    const handleDecrement = () => dispatch(decrement());
    const handleIncrementByAmount = () => dispatch(incrementByAmount(value))
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => setValue(prev => Number(event.target.value))

    return (
        <div data-testid="counter-page" className={style.container}>
            <div data-testid="value" className={style.value}>
                <span>value: </span>
                {counterValue}
                </div>
            <button data-testid="increment-button" onClick={handleIncrement}>{INCREMENT}</button>
            <button data-testid="decrement-button" onClick={handleDecrement}>{DECREMENT}</button>
            <button data-testid="increment-by-amount-button" onClick={handleIncrementByAmount}>{INCREMENT_BY_AMOUNT}</button>
            <input  data-testid="input-range" type="range" min={0} max={100} value={value} onChange={handleChange} />
            <div  data-testid="value-range" className={style.rangeValue}>{value}</div>
        </div>
    )
};

export default CounterPage;
