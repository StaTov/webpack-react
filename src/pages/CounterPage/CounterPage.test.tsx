import '@testing-library/jest-dom';
import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "src/utils/test-utils/renderWithProviders"
import userEvent from '@testing-library/user-event'
import CounterPage from './CounterPage';

const initialState = {
    value: 0
}

describe("<CounterPage/> test", () => {
    test("correct increment handle", async () => {
        renderWithProviders(<CounterPage />, {
            preloadedState: {
                counter: initialState
            }
        })
        let value = screen.getByTestId('value')
        expect(value).toHaveTextContent('0')
        const button = screen.getByTestId('increment-button')
        await userEvent.click(button)
        expect(value).toHaveTextContent('1')
    })
    test("correct decrement handle", async () => {
        renderWithProviders(<CounterPage />, {
            preloadedState: {
                counter: initialState
            }
        })
        let value = screen.getByTestId('value')
        expect(value).toHaveTextContent('0')
        const button = screen.getByTestId('decrement-button')
        await userEvent.click(button)
        expect(value).toHaveTextContent('-1')
    })
    test("correct incrementByAmount handle", async () => {
        renderWithProviders(<CounterPage />, {
            preloadedState: {
                counter: initialState
            }
        })
        let inputRange = screen.getByTestId('input-range')
        fireEvent.change(inputRange, { target: { value: "22" } })
        let value = screen.getByTestId('value')
        expect(value).toHaveTextContent('0')
        const button = screen.getByTestId('increment-by-amount-button')
        await userEvent.click(button)
        expect(value).toHaveTextContent('22')
    })
    test("correct snapshot", () => {
        const page = renderWithProviders(<CounterPage />);
        expect(page).toMatchSnapshot()
    })

})