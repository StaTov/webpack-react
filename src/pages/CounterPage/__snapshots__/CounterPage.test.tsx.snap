// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CounterPage/> test correct snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="counter-page"
      >
        <div
          class="value"
          data-testid="value"
        >
          <span>
            value: 
          </span>
          0
        </div>
        <button
          data-testid="increment-button"
        >
          increment
        </button>
        <button
          data-testid="decrement-button"
        >
          decrement
        </button>
        <button
          data-testid="increment-by-amount-button"
        >
          increment by amount
        </button>
        <input
          data-testid="input-range"
          max="100"
          min="0"
          type="range"
          value="0"
        />
        <div
          class="rangeValue"
          data-testid="value-range"
        >
          0
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="counter-page"
    >
      <div
        class="value"
        data-testid="value"
      >
        <span>
          value: 
        </span>
        0
      </div>
      <button
        data-testid="increment-button"
      >
        increment
      </button>
      <button
        data-testid="decrement-button"
      >
        decrement
      </button>
      <button
        data-testid="increment-by-amount-button"
      >
        increment by amount
      </button>
      <input
        data-testid="input-range"
        max="100"
        min="0"
        type="range"
        value="0"
      />
      <div
        class="rangeValue"
        data-testid="value-range"
      >
        0
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
